roles/kubernetes-apps/ansible/templates/nodelocaldns-config.yml.j2

```diff
@@ -0,0 +1,55 @@
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: nodelocaldns
+  namespace: kube-system
+  labels:
+    addonmanager.kubernetes.io/mode: EnsureExists
+
+data:
+  Corefile: |
+    {{ dns_domain }}:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind {{ local_dns_ip }}
+        forward . {{ clusterIP }} {
+                force_tcp
+        }
+        prometheus :9253
+        health {{ local_dns_ip }}:8080
+        }
+    in-addr.arpa:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind {{ local_dns_ip }}
+        forward . {{ clusterIP }} {
+                force_tcp
+        }
+        prometheus :9253
+        }
+    ip6.arpa:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind {{ local_dns_ip }}
+        forward . {{ clusterIP }} {
+                force_tcp
+        }
+        prometheus :9253
+        }
+    .:53 {
+        errors
+        cache 30
+        reload
+        loop
+        bind {{ local_dns_ip }}
+        forward . /etc/resolv.conf {
+                force_tcp
```

Comment:
Why force TCP?

Reply:
Extract from the KEPS document:


`Connections from local caching agent to kube-dns can be upgraded to TCP. TCP conntrack entries will be removed on connection close in contrast with UDP entries that have to timeout (default nf_conntrack_udp_timeout is 30 seconds)`

`Upgrading DNS queries from UDP to TCP would reduce tail latency attributed to dropped UDP packets and DNS timeouts usually up to 30s (3 retries + 10s timeout). Since the nodelocal cache listens for UDP DNS queries, applications don't need to be changed.`